File: src/lib/neuralModel.ts
export interface NormalizationStats {
  mean: number[];
  std: number[];
}

export interface RegressionMetrics {
  mae: number;
  rmse: number;
  r2: number;
}

export interface ModelEvaluation {
  purity: RegressionMetrics;
  yield: RegressionMetrics;
  combinedMae: number;
  combinedRmse: number;
}

export interface TrainedPeptideNetwork {
  version: string;
  activation: 'tanh';
  featureNames: string[];
  hiddenSize: number;
  inputNormalization: NormalizationStats;
  targetNormalization: NormalizationStats;
  weights: {
    W1: number[][];
    b1: number[];
    W2: number[][];
    b2: number[];
  };
  training: {
    datasetSeed: number;
    splitSeed: number;
    scenarioCounts: Record<string, number>;
    splitSizes: {
      train: number;
      validation: number;
      test: number;
    };
    bestConfig: {
      hiddenSize: number;
      learningRate: number;
      l2: number;
      maxEpochs: number;
      patience: number;
      bestEpoch: number;
    };
    validation: ModelEvaluation;
    test: ModelEvaluation;
  };
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

export function normalizeVector(values: number[], stats: NormalizationStats): number[] {
  return values.map((value, index) => (value - stats.mean[index]) / stats.std[index]);
}

export function denormalizeVector(values: number[], stats: NormalizationStats): number[] {
  return values.map((value, index) => (value * stats.std[index]) + stats.mean[index]);
}

export function forwardNormalized(model: TrainedPeptideNetwork, normalizedFeatures: number[]): number[] {
  const hidden = model.weights.b1.map((bias, rowIndex) => {
    let sum = bias;

    for (let columnIndex = 0; columnIndex < normalizedFeatures.length; columnIndex += 1) {
      sum += model.weights.W1[rowIndex][columnIndex] * normalizedFeatures[columnIndex];
    }

    return Math.tanh(sum);
  });

  return model.weights.b2.map((bias, outputIndex) => {
    let sum = bias;

    for (let hiddenIndex = 0; hiddenIndex < hidden.length; hiddenIndex += 1) {
      sum += model.weights.W2[outputIndex][hiddenIndex] * hidden[hiddenIndex];
    }

    return sum;
  });
}

export function predictTargets(model: TrainedPeptideNetwork, rawFeatures: number[]): { purity: number; yield: number } {
  const normalizedInput = normalizeVector(rawFeatures, model.inputNormalization);
  const normalizedOutput = forwardNormalized(model, normalizedInput);
  const [purity, yieldValue] = denormalizeVector(normalizedOutput, model.targetNormalization);

  return {
    purity: clamp(purity, 10, 99.5),
    yield: clamp(yieldValue, 5, 95)
  };
}

File: python/neural_model.py
from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any, Dict, List


def clamp(value: float, min_value: float, max_value: float) -> float:
    return min(max_value, max(min_value, value))


def normalize_vector(values: List[float], stats: Dict[str, List[float]]) -> List[float]:
    return [
        (value - stats["mean"][index]) / stats["std"][index]
        for index, value in enumerate(values)
    ]


def denormalize_vector(values: List[float], stats: Dict[str, List[float]]) -> List[float]:
    return [
        (value * stats["std"][index]) + stats["mean"][index]
        for index, value in enumerate(values)
    ]


def forward_normalized(model: Dict[str, Any], normalized_features: List[float]) -> List[float]:
    hidden = []
    for row_index, bias in enumerate(model["weights"]["b1"]):
        total = bias
        for column_index, feature in enumerate(normalized_features):
            total += model["weights"]["W1"][row_index][column_index] * feature
        hidden.append(math.tanh(total))

    outputs = []
    for output_index, bias in enumerate(model["weights"]["b2"]):
        total = bias
        for hidden_index, hidden_value in enumerate(hidden):
            total += model["weights"]["W2"][output_index][hidden_index] * hidden_value
        outputs.append(total)
    return outputs


def predict_targets(model: Dict[str, Any], raw_features: List[float]) -> Dict[str, float]:
    normalized_input = normalize_vector(raw_features, model["inputNormalization"])
    normalized_output = forward_normalized(model, normalized_input)
    purity, yield_value = denormalize_vector(normalized_output, model["targetNormalization"])
    return {
        "purity": clamp(purity, 10.0, 99.5),
        "yield": clamp(yield_value, 5.0, 95.0),
    }


def load_model(model_path: Path | None = None) -> Dict[str, Any]:
    resolved_path = model_path or (Path(__file__).resolve().parents[1] / "artifacts" / "peptide-model.json")
    with resolved_path.open("r", encoding="utf-8") as handle:
        return json.load(handle)
